File: civicpatch.org/src/schemas/scrape_settings.py
"""Request bodies for the cadence and budget forms.

Two requests behind one form: cadence and caps are written separately so saving one cannot
silently clear the other, which a whole-row upsert would.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class SetCadenceRequest(BaseModel):
    """`None` for cadence_days means manual: no schedule, and this state's candidates never
    drain on their own."""

    cadence_days: Optional[int] = Field(default=None, gt=0)
    # Which day the cadence lands on. Sep 1 with a 30-day cadence gives Sep 1, Oct 1,
    # Nov 1 — and Aug 2 before that, because the anchor picks the day in the cycle rather
    # than the day it starts.
    cadence_anchor: Optional[date] = None


class SetCapsRequest(BaseModel):
    """`None` means no cap at that scope, which is not the same as `0` — zero is a real setting
    meaning spend nothing, and it is what makes a cap usable as a stop switch."""

    pipeline_run_cap_usd: Optional[Decimal] = Field(default=None, ge=0)
    monthly_cap_usd: Optional[Decimal] = Field(default=None, ge=0)


class SetGlobalCapRequest(BaseModel):
    """The ceiling for every state together — a shared one, not an allocation:
    the per-state caps may sum past it, and the UI shows that rather than refusing it."""

    monthly_cap_usd: Optional[Decimal] = Field(default=None, ge=0)


class StateScrapePanel(BaseModel):
    """Everything the state's settings block shows, in one response.

    One call, not five: the block renders as a unit and five requests would let it paint in
    pieces, each arriving to a slightly different moment — the spend from one instant against a
    cap read at another.
    """

    state: str
    cadence_days: Optional[int] = None
    cadence_anchor: Optional[date] = None
    # Computed from the cadence, not read back from Temporal: the arithmetic is the same, and
    # asking would be a round-trip per state that is also wrong while a schedule is unconverged.
    # None means manual, which has no next run.
    next_run_at: Optional[datetime] = None

    pipeline_run_cap_usd: Optional[Decimal] = None
    monthly_cap_usd: Optional[Decimal] = None
    global_monthly_cap_usd: Optional[Decimal] = None

    # Month to date. Zero is honest here, unlike the spend column on the changesets page: this
    # is a figure measured against a ceiling, not a cost being reported.
    spent_this_month_usd: Decimal
    global_spent_this_month_usd: Decimal
    # Which cap is already reached, if either — `state_month`, `global_month`, or None.
    cap_reached: Optional[str] = None

    # Runs that stopped at their per-run ceiling this month. One is noise; a third of the
    # state's runs means the cap is set below what its pages cost.
    cost_cap_hits_this_month: int
    # What the next pass would pick up, right now.
    candidates_due: int


class GlobalScrapePanel(BaseModel):
    monthly_cap_usd: Optional[Decimal] = None
    spent_this_month_usd: Decimal
    # SUM of every state's monthly cap. May exceed the cap above; the block says so.
    state_monthly_caps_usd: Decimal

File: civicpatch.org/src/frontend/pages/changeset-summaries-page/scrape-settings.ts
// The cadence and budget block, and the pure formatting it needs.

import { formatUsd } from "./spend.js";

export interface StateScrapePanel {
  state: string;
  cadence_days: number | null;
  cadence_anchor: string | null;
  next_run_at: string | null;
  pipeline_run_cap_usd: string | null;
  monthly_cap_usd: string | null;
  global_monthly_cap_usd: string | null;
  spent_this_month_usd: string;
  global_spent_this_month_usd: string;
  cap_reached: string | null;
  cost_cap_hits_this_month: number;
  candidates_due: number;
}

export const MANUAL = "manual";

export function describeCadence(panel: StateScrapePanel): string {
  if (panel.cadence_days === null) return MANUAL;
  const days = panel.cadence_days;
  return days === 1 ? "daily" : `every ${days} days`;
}

// Sep 1 at 30 days gives Sep 1, Oct 1, Nov 1 — and Aug 2 before that. Not a start date.
export function describeAnchor(panel: StateScrapePanel): string {
  return panel.cadence_anchor ? `landing on ${panel.cadence_anchor}` : "";
}

export function describeNextRun(nextRunAt: string | null, now: Date): string {
  if (nextRunAt === null) return "no schedule";
  const days = Math.round((new Date(nextRunAt).getTime() - now.getTime()) / 86_400_000);
  if (days <= 0) return "due now";
  return days === 1 ? "in 1 day" : `in ${days} days`;
}

// Null cap reads as "no ceiling", never $0.00 — $0 is a real setting meaning spend nothing.
export function describeBudget(spent: string, cap: string | null): string {
  const spentText = formatUsd(spent);
  return cap === null ? `${spentText} spent, no cap` : `${spentText} of ${formatUsd(cap)}`;
}

export interface MonthlyEstimate {
  passes_per_month: number;
  per_run_usd: string;
  monthly_usd: string;
  over_cap: boolean;
}

// Display only: cost per run varies with page count and chunking, so this says "about" and
// enforcement counts real spend instead.
export function estimateMonthlyCost(
  cadenceDays: number | null,
  candidates: number,
  perRunCapUsd: string | null,
  monthlyCapUsd: string | null = null,
  fallbackPerRunUsd = "0.05",
): MonthlyEstimate | null {
  if (cadenceDays === null || candidates === 0) return null;
  const perRun = Number(perRunCapUsd ?? fallbackPerRunUsd);
  const passes = Math.max(1, Math.round(30 / cadenceDays));
  const monthly = passes * candidates * perRun;
  return {
    passes_per_month: passes,
    per_run_usd: String(perRun),
    monthly_usd: String(monthly),
    over_cap: monthlyCapUsd !== null && monthly > Number(monthlyCapUsd),
  };
}

export interface GlobalScrapePanel {
  monthly_cap_usd: string | null;
  spent_this_month_usd: string;
  state_monthly_caps_usd: string;
}

// State caps may add up past the global one: they are ceilings, not reservations.
export function describeStateCaps(panel: GlobalScrapePanel): string {
  const caps = `${formatUsd(panel.state_monthly_caps_usd)} in state caps`;
  if (panel.monthly_cap_usd === null) return caps;
  return Number(panel.state_monthly_caps_usd) > Number(panel.monthly_cap_usd)
    ? `${caps}, over the cap`
    : caps;
}
